src/v/cluster/topics_frontend.cc

```diff
@@ -1033,6 +1037,90 @@ ss::future<std::error_code> topics_frontend::change_replication_factor(
       _stm, _features, _as, std::move(cmd), timeout);
 }
 
+ss::future<topics_frontend::health_monitor_info>
+topics_frontend::get_health_info(
+  model::topic_namespace topic, int32_t partition_count) const {
+    health_monitor_info info;
+
+    partitions_filter::partitions_set_t parititon_set;
+    for (auto i = 0; i < partition_count; ++i) {
+        parititon_set.emplace(model::partition_id(i));
+    }
+
+    partitions_filter::topic_map_t topic_map;
+    topic_map.emplace(topic.tp, std::move(parititon_set));
+
+    partitions_filter partitions_for_report;
+    partitions_for_report.namespaces.emplace(topic.ns, std::move(topic_map));
+
+    node_report_filter filter;
+    filter.ntp_filters = std::move(partitions_for_report);
+
+    auto health_report = co_await _hm_frontend.local().get_cluster_health(
+      cluster_report_filter{.node_report_filter = filter},
+      force_refresh::no,
+      model::timeout_clock::now() + _get_health_report_timeout);
+
+    if (!health_report) {
+        vlog(
+          clusterlog.info,
+          "unable to get health report - {}",
+          health_report.error().message());
+        co_return info;
+    }
+
+    for (const auto& node_report : health_report.value().node_reports) {
+        uint64_t total = 0;
+        uint64_t free = 0;
+        for (const auto& disk : node_report.local_state.disks) {
+            total += disk.total;
+            free += disk.free;
+        }
+        info.node_disk_reports.emplace(
+          node_report.id, node_disk_space(node_report.id, total, total - free));
+    }
+
+    for (const auto& node_report : health_report.value().node_reports) {
+        for (const auto& tp_ns : node_report.topics) {
+            for (const auto& partition : tp_ns.partitions) {
+                info.ntp_sizes[partition.id] = partition.size_bytes;
```

Comment:
This loop may be a reactor stall on topics with 10,000+ partitions.

Also `ntp_sizes` needs to be reserved to the expected size before appending, to avoid repeated allocations.